### PR TITLE
feat(perf) Convert empty tag conditions to use tags.key

### PR DESCRIPTION
**Pull request by evanh**
Sentry is building tag queries in the form `ifNull(tags[x], '') =/!= ''`. These
checks can be rewritten to use `has(tags.key, x)` to be functionally the same
and potentially see a performance improvement as `tags.key` has a bloomfilter.

### CODE REVIEW

File: snuba/datasets/entities/discover.py [hunk 1]
```diff
@@ -460,6 +463,7 @@ def get_query_processors(self) -> Sequence[QueryProcessor]:
             TagsExpanderProcessor(),
             BasicFunctionsProcessor(),
             ProjectRateLimiterProcessor(project_column="project_id"),
+            EmptyTagConditionProcessor(),
```
Comment (fpacifici) [inline snuba/datasets/entities/discover.py:466]: Any reason for adding this processor to the logical layer? the existence of a bloomfilter index on the tags.key column is specific to the storage (not guaranteed to exist on a specific storage), so it seems an optimization to run on the clickhouse layer.

Though we are actually asking clients to apply that optimization on their own as well, so I guess ok?
If you decide to move to the clickhouse layer there is already a matcher for resolved tags:
https://github.com/getsentry/snuba/blob/master/snuba/clickhouse/translators/snuba/mappers.py#L235-L256

File: snuba/query/matchers.py [hunk 2]
```diff
@@ -349,4 +350,36 @@ def match(self, node: AnyType) -> Optional[MatchResult]:
         return result
 
 
+@dataclass(frozen=True)
+class SubscriptableReference(Pattern[SubscriptableReferenceExpr]):
+    """
+    Matches a SubscriptableReference in the AST expression.
+    It works like the Column Pattern. if alias, column_name and key arguments
```
Comment (fpacifici) [inline snuba/query/matchers.py:357]: Alias ? (there are no alias in the matchers, we ignore them)

File: snuba/query/processors/empty_tag_condition_processor.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+                replacement = FunctionCall(
+                    exp.alias,
+                    "has",
+                    (Column(None, lhs.column.table_name, "tags.key"), lhs.key),
+                )
+
+                if exp.function_name == ConditionFunctions.EQ:
+                    replacement = FunctionCall(exp.alias, "not", (replacement,))
+
+                if settings.TESTING or random.random() < 0.5:
```
Comment (fpacifici) [inline snuba/query/processors/empty_tag_condition_processor.py:55]: You sure you do not want to have this in the config ?

File: snuba/query/processors/empty_tag_condition_processor.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+                    query.add_experiment("empty-string-tag-condition", "true")
+                    return replacement
+                else:
+                    query.add_experiment("empty-string-tag-condition", "false")
+
+            return exp
+
+        condition = query.get_condition()
+        if condition is not None:
+            query.set_ast_condition(condition.transform(process_condition))
```
Comment (fpacifici) [inline snuba/query/processors/empty_tag_condition_processor.py:65]: I think this can be in the having clause as well.

**Review by fpacifici — APPROVED** [review]

File: snuba/datasets/entities/discover.py (hunk 1 above)
Comment (evanh) [reply snuba/datasets/entities/discover.py:466]: At the moment we're not asking client to add that optimization. This is for internal testing to see if it even helps. I'll move this to the storage layer, that does make more sense.

File: snuba/query/matchers.py (hunk 2 above)
Comment (evanh) [reply snuba/query/matchers.py:357]: I copied this from FunctionCall, I'll change it.

File: snuba/query/processors/empty_tag_condition_processor.py (hunk 4 above)
Comment (evanh) [reply snuba/query/processors/empty_tag_condition_processor.py:65]: Can you elaborate? I can't think of an aggregate that has this condition inside it.

Comment (fpacifici) [reply snuba/query/processors/empty_tag_condition_processor.py:65]: It was just for completeness.

**Pull request merged**